pkg/controllers/management/clusterdeploy/clusterdeploy.go

```diff
@@ -209,6 +227,65 @@ func (cd *clusterDeploy) deployAgent(cluster *v3.Cluster) error {
 	return err
 }
 
+func (cd *clusterDeploy) validateAgent(cluster *v3.Cluster) error {
+	if cluster.Status.AgentImage == "" {
+		//if empty we're waiting for something to deploy, valid until proven otherwise.
+		return nil
+	}
+
+	//check if the agent downstream is the same as the one locally in the crd
+	var na, ca string
+	var ok bool
+	var err error
+	if na, ok = nodeAgentImages[cluster.UID]; !ok {
+		na, err = cd.getNodeAgentImage(cluster)
+		if err != nil {
+			return err
+		}
+		nodeMutex.Lock()
+		nodeAgentImages[cluster.UID] = na
+		nodeMutex.Unlock()
+	}
+
+	if ca, ok = clusterAgentImages[cluster.UID]; !ok {
+		ca, err = cd.getClusterAgentImage(cluster)
+		if err != nil {
+			return err
+		}
+		clusterMutex.Lock()
+		clusterAgentImages[cluster.UID] = ca
+		clusterMutex.Unlock()
+	}
+
+	if cluster.Status.AgentImageOverride != "" {
+		if cluster.Status.AgentImageOverride == cluster.Spec.DesiredAgentImage {
+			return nil
+		}
+		if cluster.Status.AgentImageOverride == cluster.Status.AgentImage {
+			return nil
+		}
+		clearAgentCaches(cluster.UID)
+		cluster.Status.AgentImage = ""
+		cluster.Spec.DesiredAgentImage = cluster.Status.AgentImageOverride
+		return fmt.Errorf("Rancher Agent Override specificied, setting desired image: %s",
```

Comment:
why do we return err here and L283 if we're already handling it?

Reply:
forces the process to save the cluster and get us back into the redeploy code (which is above this code) on the second call to doSync

Reply:
This isn't an error and should not be handled like one.